service/src/main/java/com/epam/esm/services/service/impl/TagServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+package com.epam.esm.services.service.impl;
+
+import com.epam.esm.persistence.util.EntityFinder;
+import com.epam.esm.persistence.dao.impl.TagDAOImpl;
+import com.epam.esm.model.entity.Tag;
+import com.epam.esm.persistence.util.TagFinder;
+import com.epam.esm.persistence.exceptions.DAOSQLException;
+import com.epam.esm.services.exceptions.BadRequestException;
+import com.epam.esm.services.exceptions.ServiceException;
+import com.epam.esm.services.exceptions.ValidationException;
+import com.epam.esm.services.constants.TagSearchParameters;
+import com.epam.esm.services.constants.TagSortingParameters;
+import com.epam.esm.services.service.TagService;
+import com.epam.esm.services.validator.EntityValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+@Service
+public class TagServiceImpl implements TagService {
+
+    private TagDAOImpl dao;
```

Comment:
Please use interfaces instead of implementation

Reply:
Done.